fastq/pairedEnd.go

```diff
@@ -121,3 +121,22 @@ func GoWriteFqPair(readOne string, readTwo string, data <-chan *PairedEnd) {
 	go WritingChan(readOne, readTwo, data, &wg)
 	wg.Wait()
 }
+
+// ReadFqBigPair will take 2 readers which will convert paired end read fastq files into a paired end FastqBig struct.
+// Note: while this function is return as a pointer, it's purpose is to be deferenced at the next function call.
+// In additon the pointers to read one and read two will also be dereference. When sending a pointer to a struct through
+// a channel, or a struct with pointers inside, memory allocated will be placed on the heap hindering performance.
+func ReadFqBigPair(readerOne *fileio.SimpleReader, readerTwo *fileio.SimpleReader) (*PairedEndBig, bool) {
```

Comment:
Does sending the value as a pointer the dereferencing it in the subsequent function actually help performance? If the value is returned as a pointer then it will go to the heap and dereferencing it won't move it to the stack right?

Reply:
In short yes, this has a dramatic effect on performance only if you are passing a struct through a channel. This about optimizing what goes through the channel. However, the downside to reading non-pointers inside a loop, you can no longer check for nil to exit the loop, so a small trick you can do is to initialize a pointer check for nil, and dereference it before you send it through the channel.

Reply:
Also, i worry that the message is starting to get cloudy. I don't want it to seem like ALL heap allocations are bad, that is a very narrow point of view. This Read function addresses hold **_long_** the memory stays on the heap, not necessarily decreasing the the heap allocation.

Reply:
I see. Would this be a fair statement: "When pointers are returned from functions they can be efficiently garbage collected, but when sent through channels they cannot be garbage collected until the main function has returned"?

Reply:
i am not sure you can put it that way. i am not saying you cannot send a pointer to a channel, but if you want to reduce your heap allocation is the easiest start.

Stack allocation requires that the lifetime and memory footprint of a variable can be determined at compile time. So when you send a pointer (or structs that contain pointers) to a channel, the GC has no idea which channel you sent the data to or when you are done with that pointer and will allocate it on the heap for the duration of the function.